First implementation

ametaireau opened the pull request:
Based on top of #1 

- [x] Session handling;
- [x] Bucket handling;
- [x] Permissions handling;
- [x] Collection handling;
- [x] Records handling;

ametaireau commented on kintoclient/__init__.py:
```diff
@@ -0,0 +1,247 @@
+        self.id = id
+        self.collection = collection
+        self.permissions = Permissions('record', permissions)
+        self.data = data
```
`Request` and `Collection` objects actually issue requests themselves, whereas the `Record` waits to be populated by the `Collection.get_record`. I'm not sure which solution I prefer:
1. Always do the request in the constructor;
2. Always do the request in the `get_xxx` method of the parent object;
3. Don't bother and let it like it is right now.

leplatrem replied:
Another possibility: make ``data`` optional everywhere, and do not issue request if provided.

Also, here there could be a ``create`` kwarg too.

ametaireau replied:
I decided to make this explicit by passing a `load` argument to all the `object` representations.

I think we should still modify the code for `Record` in order to have it behave like the other ones.

In more details, it should have a `create` argument like you mentionned and have it responsible for operations on it;

Currently the collection has the responsability to issue requests about records, and it handles as well the URIs of the resource, but it should be the `Resource` responsability.

I will update the PR to reflect this shortly.

leplatrem commented on README.rst:
```diff
@@ -1,6 +1,220 @@
-Python library to interact with a `Kinto <https://kinto.readthedocs.org>`_
-HTTP server.
+Kinto is a service allowing you to store and synchronize arbitrary data,
+attached to an user account. It's primary interface is HTTP.
```
to a user account [...] Its primary

leplatrem commented on README.rst:
```diff
@@ -1,6 +1,220 @@
+attached to an user account. It's primary interface is HTTP.
+
+`kinto-client` is a python library which aims to ease the interaction with
+kinto for clients. `A project with related goals is
```
nit:  *Kinto* (capitalized, italic?)

leplatrem commented on README.rst:
```diff
@@ -1,6 +1,220 @@
+
+    The approach this library takes is different from the one described in the
+    `Cliquetis <https://github.com/mozilla-services/cliquetis>`_ javascript
+    client: Operations are always performed directly on the server, and no
```
It would remove the beginning and just start with "Operations are..."

leplatrem commented on README.rst:
```diff
@@ -1,6 +1,220 @@
+  refresh mechanism. If you want to be sure you have the last data available,
+  issue another call.
+
+Here are is an overview of what the API looks like:
```
nit: Here is

leplatrem commented on README.rst:
```diff
@@ -1,6 +1,220 @@
+Records
+=======
+
+Records can be retrieved from and saved to buckets.
```
collections ?

leplatrem commented on kintoclient/__init__.py:
```diff
@@ -0,0 +1,254 @@
+
+DEFAULT_SERVER_URL = 'https://kinto.dev.mozaws.net/v1'
+
+# XXX rename to 'objects'?
```
+1 to rename to objects (see http://kinto.readthedocs.org/en/latest/permissions.html)

leplatrem commented on kintoclient/__init__.py:
```diff
@@ -0,0 +1,254 @@
+        self.session.request('delete', uri)
+
+    def create_group(self, name, members):
+        pass
```
raise NotImplementedError ?

leplatrem commented on kintoclient/__init__.py:
```diff
@@ -0,0 +1,254 @@
+        uri = "%s/%s" % (self.uri, 'collections')
+        resp, _ = self.session.request('get', uri)
+
+        return [collection['id'] for collection in resp['data']]
```
nit: ``'id'`` could be a constant ID_FIELD

leplatrem commented on kintoclient/__init__.py:
```diff
@@ -0,0 +1,254 @@
+
+    def delete_collection(self, name):
+        uri = self._get_collection_uri(name)
+        self.session.request('delete', uri)
```
Since the server returns content on delete, I believe the client could also return it

leplatrem commented on kintoclient/__init__.py:
```diff
@@ -0,0 +1,254 @@
+
+    def delete_record(self, id):
+        record_uri = self._get_record_uri(id)
+        resp, _ = self.session.request('delete', record_uri)
```
same remark about returned value for delete

leplatrem commented on kintoclient/__init__.py:
```diff
@@ -0,0 +1,254 @@
+        self.data = info['data']
+        self.permissions = Permissions(
+            container='bucket',
+            permissions=info.get('permissions'))
```
Elsewhere you used ``['permissions']``

leplatrem commented on kintoclient/tests/test_client.py:
```diff
@@ -0,0 +1,367 @@
+    def test_collections_can_be_created_from_buckets(self, collection_mock):
+        bucket = Bucket('testbucket', session=self.session)
+        bucket.create_collection('mycollection')
+        collection_mock.assert_called_with(
```
I believe you could just assert instanceof and ``collection.id == 'mycollection'``

ametaireau replied:
Not really, here I'm also testing that the collection is created, and that the correct arguments are passed.

leplatrem commented on kintoclient/tests/test_client.py:
```diff
@@ -0,0 +1,367 @@
+
+class PermissionsTests(TestCase):
+
+    def test_should_throw_on_invalid_container(self):
```
nit: fails with...

leplatrem commented on kintoclient/tests/test_client.py:
```diff
@@ -0,0 +1,367 @@
+    def test_should_throw_on_invalid_container(self):
+        self.assertRaises(AttributeError, Permissions, 'unknown_container')
+
+    def test_should_not_throw_on_valid_container(self):
```
nitto

leplatrem commented:
It looks very good !

I have one major remark : now that there is an implementation, the dreamcode should be removed from the README. Otherwise there is a big risk that some dream features remain described without being implemented :)

I suggest we open an issue to add the setup of an integration task on Travis like for kinto.js, since the unit tests do not garantee compability with the server.

ametaireau commented:
Okay, here is the current state of this pull request:

- I've added functional tests for everything (they can be run with `make functional`);
- I've updated the implementation along the way, and then fixed the test suite;
- As a result, the coverage (was 98% before the changes) is now around 70%, and needs to be fixed. 

Here are my notes on what's next to do for this PR:

- Add tests for the passing of permissions on bucket creation;
- Add tests for the new exception handling mechanism;
- Add tests for bucket deletion;
- Add tests for the `__repr__` method for Permissions;
- Add other `__repr__` methods for other objects;
- Add tests for `permissions.as_dict` called only when the as_dict method exists.
- Add tests for the collection deletion;
- Add tests for record deletion;
- Add tests about the load parameter (Bucket, Collection, Record);
- Check that load=False and create=True cannot be specified at the same time;
- Add a test to experience bucket creation when the `bucket` parameter is a string.

kumar303 commented on README.rst:
```diff
@@ -181,10 +180,10 @@ Sending requests in batch
+operations have been done offline and the server needs a refresh.
 
-Batch operations can be done using a python context manager (the `with`
+Batch operations can be done using a Python context manager (the `with`
```
nice! Cool way to do batching

kumar303 commented on README.rst:
```diff
@@ -195,8 +194,10 @@ bucket. It is possible to pass the session to the constructor of the `Bucket`.
-    session.commit()
+    my_bucket = Bucket('personal', session=session, create=True)
+    for collection in range(5):
+        my_bucket.create_collection("toto-%s" % collection, create=True)
```
wouldn't `create_collection()` imply `create=True` ?

ametaireau replied:
yeah, I actually need to update this README :)

kumar303 commented on dev-requirements.txt:
```diff
@@ -0,0 +1,7 @@
+pytest
```
I find it's always useful to hint at the version you were using like `pytest>=X.X.X` because with open ended requirements, it's always possible that something will break your code in the future (probably for someone else :)). If you at least have a hint to which version *used to work* then someone can deduce the problem easier.

Natim replied:
What we chose to do in Python is to ping the version at release time but not in master.
See: https://github.com/Kinto/kinto/blob/master/setup.py#L18-L21 vs https://github.com/Kinto/kinto/blob/1.3.1/setup.py#L18-L21

kumar303 replied:
Ah, that's pretty cool. That looks like it only pins production requirements at release time though. I find it equally as important to pin dev requirements because this is the environment that breaks more often for other developers.

kumar303 commented on kintoclient/__init__.py:
```diff
@@ -0,0 +1,344 @@
+                  .setdefault('Content-Type', 'application/json')
+            kwargs.setdefault('data', json.dumps(payload))
+        resp = requests.request(method, actual_url, **kwargs)
+        if not (200 <= resp.status_code < 400):
```
it might be more straight forward to call and catch [resp.raise_for_status()](http://docs.python-requests.org/en/latest/api/#requests.Response.raise_for_status)

ametaireau replied:
I was actually doing that before but since I'm raising another exception afterwards, but it's almost doing the same thing.

ametaireau replied:
Actually `raise_for_status` [doesn't check the same way](https://github.com/kennethreitz/requests/blob/f5dacf84468ab7e0631cc61a3f1431a32e3e143c/requests/models.py#L825-L837) and an hypothetical status code which would be less that 200 wouldn't raise.

It [makes sense](https://pretty-rfc.herokuapp.com/RFC2616#status.1xx) to have this in requests, but in Kinto.py, we expect to never see status code less that 200.

kumar303 commented on kintoclient/__init__.py:
```diff
@@ -0,0 +1,344 @@
+            bucket = Bucket(bucket, session=self.session, load=False)
+        self.bucket = bucket
+        self.name = name
+        self.uri = "%s/collections/%s" % (self.bucket.uri, self.name)
```
shouldn't the name be urlencoded here?

Natim replied:
At the moment there are no authorized characters that needs to be urlencoded.
See: https://github.com/mozilla-services/cliquet/blob/master/cliquet/storage/generators.py#L14

But if we want to be future proof on that we may want to do so.

ametaireau replied:
yep, good catch!

kumar303 commented on kintoclient/tests/config/kinto.ini:
```diff
@@ -0,0 +1,49 @@
+kinto.flush_endpoint_enabled = true
+
+[server:main]
+use = egg:waitress#main
```
waitress? :eyes: :heart_eyes:

Natim replied:
It is the wsgi backend used to run the test server.

kumar303 commented on kintoclient/tests/test_client.py:
```diff
@@ -0,0 +1,541 @@
+
+        try:
+            Bucket('test', session=self.session)
+        except BucketNotFound as e:
```
in unittest you can do [assertRaises()](https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertRaises):
````
with self.assertRaises(BucketNotFound) as e:
    Bucket('test', session=self.session)
self.assertEquals(e.response, exception.response)
...
````

ametaireau replied:
Thanks for pointing that out.

That's reassuring to see that we make the same mistake, though, the context manager is actually a context manager and not the exception itself, so you'll need to do something like that:

```python
with self.assertRaises(BucketNotFound) as cm:
    Bucket('test', session=self.session)

e = cm.exception
self.assertEquals(e.response, exception.response)
```

kumar303 replied:
ah, right. I also find it confusing that you can't make an assertion about the exception while still in `with` block.

kumar303 commented on kintoclient/tests/test_client.py:
```diff
@@ -0,0 +1,541 @@
+            data=json.dumps({'data': {'foo': 'bar'}}),
+            headers={'Content-Type': 'application/json'})
+
+    @mock.patch('kintoclient.requests')
```
since this is used so much you could also put it in setUp:
````
def setUp(self):
    p = mock.patch('kintoclient.requests')
    self.requests_mock = p.start()
    self.addCleanup(p.stop)
````

ametaireau replied:
Thanks :-)

kumar303 commented on setup.py:
```diff
@@ -0,0 +1,34 @@
+here = os.path.abspath(os.path.dirname(__file__))
+
+with codecs.open(os.path.join(here, 'README.rst'), encoding='utf-8') as f:
+    README = f.read()
```
if you're going to read this from your setup.py you have to also add `README.rst` to [MANIFEST.in](https://docs.python.org/2/distutils/sourcedist.html#the-manifest-in-template)

Natim replied:
:+1:

kumar303 commented:
API looks cool to me. I just commented on some minor things.

Natim commented on tox.ini:
```diff
@@ -0,0 +1,14 @@
+[tox]
+envlist = py27,flake8
```
Could we add py34 here?

ametaireau replied:
not just yet I believe, but we should.

Natim commented:
We should probably merge this and continue to improve it in other PR.
